simd/src/Kokkos_SIMD_Common_Math.hpp

```diff
@@ -57,17 +64,47 @@ hmax(const_where_expression<basic_simd_mask<T, Abi>, basic_simd<T, Abi>> const&
   }
   return result;
 }
+#endif
 
-template <class T, class Abi>
-[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T reduce(
-    const_where_expression<basic_simd_mask<T, Abi>, basic_simd<T, Abi>> const&
-        x,
-    T, std::plus<>) {
-  auto const& v = x.impl_get_value();
-  auto const& m = x.impl_get_mask();
-  auto result   = Kokkos::reduction_identity<T>::sum();
+template <
+    typename T, typename Abi,
+    std::enable_if_t<!std::is_same_v<Abi, simd_abi::scalar>, bool> = false>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+reduce_min(basic_simd<T, Abi> const& v,
+           typename basic_simd<T, Abi>::mask_type const& m) {
+  auto result = Kokkos::reduction_identity<T>::min();
+  for (std::size_t i = 0; i < v.size(); ++i) {
+    if (m[i]) result = Kokkos::min(result, v[i]);
+  }
+  return result;
+}
+
+template <
+    class T, class Abi,
+    std::enable_if_t<!std::is_same_v<Abi, simd_abi::scalar>, bool> = false>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+reduce_max(basic_simd<T, Abi> const& v,
+           typename basic_simd<T, Abi>::mask_type const& m) {
+  auto result = Kokkos::reduction_identity<T>::max();
   for (std::size_t i = 0; i < v.size(); ++i) {
-    if (m[i]) result += v[i];
+    if (m[i]) result = Kokkos::max(result, v[i]);
+  }
+  return result;
+}
+
+template <
+    class T, class Abi, class BinaryOperation = std::plus<>,
+    std::enable_if_t<!std::is_same_v<Abi, simd_abi::scalar>, bool> = false>
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION T
+reduce(basic_simd<T, Abi> const& v,
+       typename basic_simd<T, Abi>::mask_type const& m, T identity,
+       BinaryOperation op = {}) {
+  if (none_of(m)) {
+    return identity;
+  }
+  auto result = v[0];
+  for (std::size_t i = 1; i < v.size(); ++i) {
+    if (m[i]) result = op(result, v[i]);
   }
```

Comment:
```suggestion
  auto result = identity;
  for (std::size_t i = 0; i < v.size(); ++i) {
    if (m[i]) result = op(result, v[i]);
  }
```

Reply:
Using `identity` wasn't entirely correct as `identity` could be a user provided arbitrary value. Modified to use a true identity element to fix this. See [d3211e1](https://github.com/kokkos/kokkos/pull/7486/commits/d3211e143d048ebd9e4e1fbc0171b112794bdb5b).